### ISSUE TITLE
[Docker] - Production images and Development environment

### ISSUE THREAD

**Issue by ceckoslab** [basicrum/backoffice#34]
This idea is still far from well defined but we need to start from somewhere.

So far I develop on my computer just fine but I am also working on SaaS version of Basic RUM and I am having some difficulties to build the proper Docker files and reuse them for production and development environment.

I need an expert who can build Docker images and understands how to build Symfony 4 development and production environment.

So far I have docker-compose.yml and related files in **docker** folder of this project.

I have 4 containers: Nginx, PHP-FPM, MySql and Composer but we may combine Nginx and PHP-FPM. I think that Composer container is not need because it was part of old experiments.

We also need the possibility to have Let's Encript for production environment.

We also have docker/db/data/basicrum_demo_2019-04-28.sql.gz which is used for import for local development. It really helps with onboarding process. 

This is what we have so far from some experimental work:

```
ARG BASEIMAGE="tianon/true"
FROM composer:latest as composer-builder

#RUN rm -rf /var/www && mkdir /var/wwww
WORKDIR /var/www

#COPY ./backoffice/composer.* /var/www/
COPY ./backoffice/* /var/www/

ARG APP_ENV=prod

RUN set -xe \
    && if [ "$APP_ENV" = "prod" ]; then export ARGS="--no-dev"; fi \
    && composer install --prefer-dist --no-scripts --no-progress --no-suggest --no-interaction --apcu-autoloader $ARGS

RUN composer dump-autoload --classmap-authoritative


FROM ${BASEIMAGE}

COPY --from=composer-builder /var/www /
```

Current focus should be building containers needed for **Monolith Basic RUM** which could be found on this diagram:
![alt Basic RUM SaaS infrastructure](https://raw.githubusercontent.com/basicrum/infrastructure/master/docs/saas-infrastructure.jpg)

**Comment by ceckoslab** [comment]
Hello @eliskovets 

Could you take a look again at this issue and check if the code in this PR https://github.com/basicrum/backoffice/pull/40 still works with the changes that were implemented during last couple of months?

Have a nice day!

**Comment by eliskovets** [comment]
Hi @ceckoslab , 
I went through the description of this issue once again and I believe it would be better to keep only local configuration in the repo, so everybody can give it a try on their local environment. 
We can prepare a separate repo with docker-compose for production environment (in case if somebody would like to run it on the top of docker swarm) as well. May be with some additional adjustability. + helm chart can be another option in case if people would like to deploy in on their K8S cluster.  

Anyway, I believe for SAAS solution the configuration is going to be different, as it should be designed for multiuser environment. For a self-hosted BasicRUM we don't need anything similar. 

What do you think?

**Comment by ceckoslab** [comment]
Hello @eliskovets 

Thanks for checking this in details! 

I agree with what you are saying and here are my comments:

- Let's keep it as you suggested about the local development and perhaps have fixed env configuration and docker compose files. This should make things simply for the core group and the new potential contributors.
- Probably we could do the production related stuff here: https://github.com/basicrum/infrastructure
- I love the idea for **helm chart** and probably it will be helpful for us when we work on the SAAS project.
- I believe that we need to give a chance to Basic RUM users to run the monolith version on production without helm because having helm could could be complex for some users.

That being said we can keep the helm and non helm code bases open sourced :)

**Comment by eliskovets** [comment]
Hi @ceckoslab .
The PR is ready for your review. 
I've configured local env with makefile. Starting it on new environment should be as simple as to run `make init` + `make demo` (in case if demo data should be preloaded).
I had to convert the DB dump with the data to data-only dump, so it should not rewrite the DB configuration like user/pass and table structure anymore. 
As a part of the PR I've also prepared a docker build and push script for tags only. We can extend it in future if we need to push some additional releases, as nightly build and so on. 

I know that this PR is not actually about what's written in the description of this ticket, but it's a first step. As a second step I would suggest you to create 2 separate repositories: 1 for swarm configuration and another one for helm:
- swarm-config
- helm-charts
There we can prepare a configuration for production like environment contains backoffice and digger part.

PS: I believe we need to move[ digger's part](https://github.com/revampix/boomerang-beacon-catcher/blob/master/app/public/excavator/digger.php) of BasicRum under this git organisation as well.

**Comment by ceckoslab** [comment]
Hello @eliskovets 

Thanks for working on this. I looked at the PR and it looks great! I will play a bit more with it today.

1. What do you think about exposing the data files of db container to the host? Do you think that this could help us with doing backups for local environment much easier?

2. I moved the digger and one more repository to Basic RUM organization:
2.1. https://github.com/basicrum/boomerang-beacon-catcher
2.2. https://github.com/basicrum/boomerang-resourcetiming-decompression-php-port

3. Do you want me to name the new repositories:

- swarm-config
- helm-charts

**Comment by ceckoslab** [comment]
Hello @NikolovI and @korzol 

This task is currently under review but once merged I believe that it will introduce breaking changes for developers who are using docker setup.

@NikolovI I am not sure how this will run under Windows because for now we tested it only under Mac OS and Ubuntu. I hope that you will help us figure out how the setup will work under Windows.

Here is something about the breaking changes: We change the username and password for the MySql database: https://github.com/basicrum/backoffice/pull/114/files#diff-072e81fe88d9b23fe00fc189279936a6R42

Also after we merge this to master you may need to delete .env file from the root of your project.

I advice you to do a backup If you have important data in your local DB.

The benefits:

For local development @eliskovets introduced a **make** file that wraps various docker commands that previously we had to type by hand: https://github.com/basicrum/backoffice/pull/114/files#diff-b67911656ef5d18c4ae36cb6741b7965R1

Here is what the make file could do for us: 

```
tsvetan@tsvetan-XPS-15-9550:~/projects/basicrum/backoffice$ make help
beacons                        Prepare beacons folders
cc                             Clean cache
demo                           Loads demo data to the DB
destroy                        Destroy local environment
docker_publish                 Publish new image to docker hub
down                           Stop local environment
init_script                    Inistall dependencies and apply migrations
init                           Initialise environment on a first start
jumpapp                        Jump to the app container
jumpdb                         Jump to db container
migrate                        Apply migrations
rebuild                        Rebuild local environment from scratch
test                           Start tests on local environment
up                             Start local environment
user                           Creates superadmin user
```

@NikolovI it's really interesting how we can run the make file on windows, how that we can figure out this together.

**Comment by ceckoslab** [comment]
Merged to master

**Comment by NikolovI** [comment]
Hello @ceckoslab ,

I've managed to get things running with Cygwin. I had to make some minor modifications in the makefile in order to test it quickly (all related to the docker_build and docker_push tasks), which doesn't matter for me at this point. I think they are related to my local environment so I'll go back to them when I need them. The project is starting up with the new setup, I'm having an issue with the demo data loading but I'm going to describe it in new issue.

Thanks to @eliskovets for the much easier setup experience :)
